### ISSUE TITLE
Add check for hermiticity in DMRG and related functions (?)

### ISSUE THREAD

**Issue by sujay-kazi** [ITensor/ITensors.jl#262]
It seems that DMRG sometimes fails to notice that a Hamiltonian is not Hermitian. For example, the (Sx)(Sz) operator on a single site is not Hermitian--its matrix representation is [0   0   0  ;  1/sqrt(2)   0   -1/sqrt(2)  ;  0   0   0]. However, when I run the code shown below:

```
using ITensors
sites = siteinds("S=1",N)
ampo = AutoMPO()
add!(ampo,1,"Sx",1,"Sz",1)
H = MPO(ampo,sites)
psi0 = randomMPS(sites)
sweeps = Sweeps(5)
maxdim!(sweeps, 10,20,100,100,200)
cutoff!(sweeps, 1E-10)
energy, psi = dmrg(H, psi0, sweeps)
```

I do not get an error. This is similarly true if I run the above code with (Sz)(Sx) (i.e. just change the 4th line out of 10 in the above code). However, if I run the above code with (Sx)(Sy), (Sy)(Sx), (Sy)(Sz), or (Sz)(Sy), the fact that the Hamiltonian is not Hermitian is caught at the beginning of DMRG. In such cases, I get an error of the form shown below:

```
ERROR: LoadError: operator does not appear to be hermitian: 0.2523186340831955 vs 0.30567776302195154
Stacktrace:
 [1] #initialize#23(::Int64, ::typeof(KrylovKit.initialize), ::KrylovKit.LanczosIterator{ProjMPO,ITensor{2},KrylovKit.ModifiedGramSchmidt2}) at C:\Users\user\.julia\packages\KrylovKit\oJ59b\src\krylov\lanczos.jl:78
 [2] #initialize at .\none:0 [inlined]
 [3] eigsolve(::ProjMPO, ::ITensor{2}, ::Int64, ::Symbol, ::KrylovKit.Lanczos{KrylovKit.ModifiedGramSchmidt2,Float64}) at C:\Users\user\.julia\packages\KrylovKit\oJ59b\src\eigsolve\lanczos.jl:10
 [4] #eigsolve#41(::Base.Iterators.Pairs{Symbol,Real,NTuple{4,Symbol},NamedTuple{(:ishermitian, :tol, :krylovdim, :maxiter),Tuple{Bool,Float64,Int64,Int64}}}, ::typeof(KrylovKit.eigsolve), ::ProjMPO, ::ITensor{2}, ::Int64, ::Symbol) at C:\Users\user\.julia\packages\KrylovKit\oJ59b\src\eigsolve\eigsolve.jl:163
 [5] (::KrylovKit.var"#kw##eigsolve")(::NamedTuple{(:ishermitian, :tol, :krylovdim, :maxiter),Tuple{Bool,Float64,Int64,Int64}}, ::typeof(KrylovKit.eigsolve), ::ProjMPO, ::ITensor{2}, ::Int64, ::Symbol) at .\none:0
 [6] macro expansion at C:\Users\user\.julia\packages\ITensors\6I0QC\src\mps\dmrg.jl:55 [inlined]
 [7] macro expansion at .\util.jl:212 [inlined]
 [8] #dmrg#242(::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::typeof(dmrg), ::MPO, ::MPS, ::Sweeps) at C:\Users\user\.julia\packages\ITensors\6I0QC\src\mps\dmrg.jl:42
 [9] dmrg at C:\Users\user\.julia\packages\ITensors\6I0QC\src\mps\dmrg.jl:8 [inlined]
 [10] top-level scope at D:\Julia Files\spin_1_model.jl:93
in expression starting at D:\Julia Files\spin_1_model.jl:28
```

I don't see why the program successfully detects a non-Hermitian Hamiltonian in some cases but not others. The only real difference I see in the examples above is that (Sx)(Sz) and (Sz)(Sx) have all real components, while the others don't. Any help with this would be greatly appreciated.

**Comment by emstoudenmire** [comment]
Hi, it’s a good question. Basically, nothing in ITensors.jl right now explicitly checks that the input Hamiltonian is Hermitian. The reason is that it’s non-trivial to devise a simple and computationally efficient check that a given MPO is Hermitian, though we could likely come up with a reasonably efficient one with some effort (e.g. computing matrix elements of H between two random MPS of small bond dimension).

Here what you are seeing is that the error is being generated from the KrylovKit code, which is an external dependency (a different library outside of ITensor, as you probably know). So it’s hard for us to say a priori why the check it does only sometimes succeeds at detecting a non-Hermitian H. Probably it’s because it doesn’t check H directly, but just looks at properties of H within the Krylov basis, which may look ok for certain non-Hermitian H’s. 

In conclusion, we could add code that checks whether H is Hermitian within ITensor itself, but we just haven’t made this a priority and it requires some non-trivial engineering to do well.

**Comment by mtfishman** [comment]
I'll also add that we plan to support non-Hermitian DMRG by adding an `ishermitian` keyword argument, where it will default to `true`. In general, in functions that support both Hermitian and non-Hermitian versions (like `eigen`), we do not have checks to make sure that the user inputs a Hermitian operator when `ishermitian=true`, since as Miles said these checks cannot always be done with trivial cost. We could consider doing some quick checks (like checking random matrix elements), but in general I think it is up to the user to ensure the correct input.

We can keep this as an open issue to possibly add checks for Hermiticity in functions that expect Hermitian inputs.
